### PR TITLE
Bitswap Quest

### PR DESCRIPTION
**Pull request by dignifiedquire**
This is still a work in progress

### TODO

- [x] Integrate bitswap
- [x] Simple block exchange test
- [x] Complex block exchange test
- [x] File exchange test
- [x] Bitswap Commands
  - [x] HTTP
    - [x] `stat`
    - [x] `wantlist`
    - ~~`unwant` will be done later~~
  - [x] CLI
    - [x] `stat`
    - [x] `wantlist`
    - ~~`unwant` will be done later~~
- [x] CI Green

### Needs the following PRs

- [x] https://github.com/ipfs/js-ipfs-bitswap/pull/12
- [x] https://github.com/ipfs/js-ipfs-block-service/pull/19
- [x] https://github.com/diasdavid/js-libp2p-swarm/pull/49
- [x] https://github.com/ipfs/js-ipfs-api/pull/264
- [x] https://github.com/ipfs/js-ipfs-bitswap/pull/10

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -14,12 +14,21 @@ script:
...
 
 before_script:
   - export DISPLAY=:99.0
   - sh -e /etc/init.d/xvfb start
 
 after_success:
   - npm run coverage-publish
+
+env:
+  - CXX=g++-4.8
```
Comment (diasdavid) [inline .travis.yml]: why is this necessary?

File: src/cli/commands/bitswap/unwant.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    }
+  },
+
+  run: (key) => {
+    utils.getIPFS((err, ipfs) => {
+      if (err) {
+        throw err
+      }
+
+      // TODO: implement
```
Comment (diasdavid) [inline src/cli/commands/bitswap/unwant.js]: Let's throw a error here saying that it is not implemented yet

File: .travis.yml (hunk 1 above)
Comment (dignifiedquire) [inline .travis.yml]: So that travis can build native modules (node 4 and greater require a c++ compiler for those) Details here: https://docs.travis-ci.com/user/languages/javascript-with-nodejs#Node.js-v4-(or-io.js-v3)-compiler-requirements

File: src/core/ipfs/go-offline.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+const debug = require('debug')
+const log = debug('core:offline')
+
+module.exports = function goOffline (self) {
+  return (cb) => {
+    self._blockS.goOffline()
+    self._bitswap.stop()
+    self.libp2p.stop((err) => {
+      if (err) {
+        log('Error trying to go offline', err)
```
Comment (diasdavid) [inline src/core/ipfs/go-offline.js]: no return, no throw?

File: src/core/ipfs/go-offline.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+'use strict'
+
+const debug = require('debug')
+const log = debug('core:offline')
+
+module.exports = function goOffline (self) {
+  return (cb) => {
+    self._blockS.goOffline()
```
Comment (diasdavid) [inline src/core/ipfs/go-offline.js]: doesn't blockS also stop bitswap?

Comment (dignifiedquire) [inline src/core/ipfs/go-offline.js]: no it just removes bitswap from the blockservice

File: src/core/ipfs/go-online.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+      }
+
+      self._bitswap = new Bitswap(
+        self._peerInfo,
+        self._libp2pNode,
+        self._repo.datastore,
+        self._peerInfoBook
+      )
+      self._bitswap.start()
+      self._blockS.goOnline(self._bitswap)
```
Comment (diasdavid) [inline src/core/ipfs/go-online.js]: I'm a bit confused why doesn't blockService start bitswap, if in order for blockService to go online, it requires bitswap to go online

File: src/core/ipfs/init.js [hunk 6]
```diff
@@ -17,7 +17,7 @@ module.exports = function init (self) {
     opts.bits = opts.bits || 2048
 
     // Pre-set config values.
-    var config = require('../../init-files/default-config.json')
+    var config = JSON.parse(fs.readFileSync(path.join(__dirname, '../../init-files/default-config.json')).toString())
```
Comment (diasdavid) [inline src/core/ipfs/init.js]: 👍

File: src/core/ipfs/go-online.js (hunk 5 above)
Comment (dignifiedquire) [inline src/core/ipfs/go-online.js]: Because the block service doesn't control bitswap, it just receives a running bitswap node.

File: src/core/ipfs/go-offline.js (hunk 3 above)
Comment (dignifiedquire) [inline src/core/ipfs/go-offline.js]: fixed

File: src/core/ipfs/init.js [hunk 7]
```diff
@@ -85,7 +85,7 @@ module.exports = function init (self) {
           throw err
         }
         const index = __dirname.lastIndexOf('/')
-        async.eachLimit(res, 10, (element, callback) => {
+        parallelLimit(res.map((element) => (callback) => {
```
Comment (diasdavid) [inline src/core/ipfs/init.js]: I wonder if we should start calling `next` the functions that are not exactly `callbacks` in the async way.

Comment (dignifiedquire) [inline src/core/ipfs/init.js]: why are they not exactly callbacks?

File: .travis.yml (hunk 1 above)
Comment (diasdavid) [inline .travis.yml]: I mean, which are the native modules that we have introduced with bitswap?

**Comment by diasdavid** [comment]
LGTM :)

File: .travis.yml (hunk 1 above)
Comment (dignifiedquire) [inline .travis.yml]: None that we really need, travis was just complaining and I wanted to fix that (those native modules are mostly used through aegir)

File: src/core/ipfs/init.js (hunk 7 above)
Comment (diasdavid) [inline src/core/ipfs/init.js]: It was just a thought :)

**Comment by dignifiedquire** [comment]
🎉  🎈 🎉  🎈 🎉  🎈 🎉  🎈 🎉  🎈 🎉  🎈 🎉  🎈 🎉  🎈

**Comment by diasdavid** [comment]
wowowowwowo :D
